### PR TITLE
Fix FP reported in #466

### LINKED ISSUE #466: `A7-1-2`: Consider `constexpr` function behaviour
### Affected rules
 - `A7-1-2`

### Description

A `constexpr` function only returns a compile-time constant when provided with compile-time constants as arguments. The query currently incorrectly considers `constexpr` functions to always return constants.

### Example

This function should only be considerd
```cpp
constexpr int add(int x, int y) {
  return x + y;
}

void test(int p) {
  int x = add(1,2); // NON_COMPLIANT - could be marked constexpr
  int y = add(1,p); // COMPLIANT - cannot be marked constexpr
```

### PR DESCRIPTION
**Pull request by rvermeulen**
## Description

Fixes #466 
## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - A7-1-2

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [x] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [x] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [ ] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [ ] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

File: cpp/autosar/src/rules/A7-1-2/VariableMissingConstexpr.ql [hunk 1]
```diff
@@ -33,6 +33,15 @@ predicate isTypeZeroInitializable(Type t) {
   t.getUnderlyingType() instanceof ArrayType
 }
 
+predicate isCompileTimeEvaluated(Call call) {
+  call.getTarget().isConstexpr() and
+  forall(Expr arg | arg = call.getAnArgument() |
```
Comment (knewbury01) [inline cpp/autosar/src/rules/A7-1-2/VariableMissingConstexpr.ql:38]: I think this logic needs another consideration

literals flowing into a var might not be sufficient

consider:
```
int a = 0;
  if (p> 0){
   a= 1;  
  }else{
    a =p+1; 
  }
    
  int x = add(a, 2); // cannot be made constexpr but currently being reported
```

if the literal comes from a constexpr var itself though then it is fine - like the following (actually also fine if a is not declared `constexpr` - but in this case it also could be):
```
  constexpr int a = 0;    
  int x = add(a, 2); // NON_COMPLIANT
```

**Review by knewbury01 — CHANGES_REQUESTED** [review]
some logic requiring reconsideration I believe

File: cpp/autosar/src/rules/A7-1-2/VariableMissingConstexpr.ql (hunk 1 above)
Comment (rvermeulen) [reply cpp/autosar/src/rules/A7-1-2/VariableMissingConstexpr.ql:38]: Good catch. I definitely need to consider more corner cases.

File: cpp/autosar/test/rules/A7-1-2/test.cpp [hunk 2]
```diff
@@ -204,4 +204,44 @@ class ExcludedCases {
...
+  int w = add1(x, 2); // NON_COMPLIANT
+  int a = add1(x);    // NON_COMPLIANT
+  int b = add1(1);    // NON_COMPLIANT
+  int c = add1(1, z); // COMPLIANT
+  int d = add1(1, z); // COMPLIANT
+  int e = add1(z);    // COMPLIANT
+  int f = add2(1);    // NON_COMPLIANT
+  int g = add2(1, 2); // NON_COMPLIANT
+  int h = add2(x, 2); // NON_COMPLIANT
+  int i = add2(x, 2); // NON_COMPLIANT
```
Comment (knewbury01) [inline cpp/autosar/test/rules/A7-1-2/test.cpp:238]: so I'll confess, I dont know the exact details of the constant propagation type logic that most compilers will use to checkup on if this can be `constexpr`... so I just added that specifier to all of the NON_COMPLIANT cases and the ones that disallow from the example compiling are : line 237,238,230,228 (maybe a few others)

additionally the readability of this section would benefit from some overhead comments that describe what each grouping exercises

**Review by knewbury01 — CHANGES_REQUESTED** [review]

File: cpp/autosar/test/rules/A7-1-2/test.cpp (hunk 2 above)
Comment (rvermeulen) [reply cpp/autosar/test/rules/A7-1-2/test.cpp]: I've added clarifications to what is supposed to be tested. The compiler errors are caused because the functions receive a compile time constant but the variable couldn't be marked `constexpr` because the intermediate variable holding the compile time constant wasn't a `constexpr`. 

Because of the resulting compilation errors I`ve opted to not present all the result to prevent confusion. It does mean that a change can result in multiple new results, but that is the trade off that has to be made.

**Review by knewbury01 — APPROVED** [review]
nice! 

I was naively thinking getting those extra testcases fairly accurate was as simple as - if another assignment (to anything) exists control flow upwards cut those out, but I definitely think that would also have not been as precise at what you implemented

very cool, think it looks good
